fix(ngcc): do not fail when multiple workers try to create the same directory

gkalpak opened the pull request:
When `ngcc` is running in parallel mode (usually when run from the command line) and the `createNewEntryPointFormats` option is set to true (e.g. via the `--create-ivy-entry-points` command line option), it can happen that two workers end up trying to create the same directory at
the same time. This can lead to a race condition, where both check for the directory existence, see that the directory does not exist and both try to create it, with the second failing due the directory's having already been created by the first one. Note that this only affects directories and not files, because `ngcc` tasks operate on different sets of files.

This commit avoids this race condition by allowing `FileSystem`'s `ensureDir()` method to not fail if one of the directories it is trying to create already exists (and is indeed a directory). This is fine for
the `ensureDir()` method, since it's purpose is to ensure that the specified directory exists. So, even if the `mkdir()` call failed (because the directory exists), `ensureDir()` has still completed its mission.

Related discussion: https://github.com/angular/angular/pull/33049#issuecomment-540485703
[FW-1635](https://angular-team.atlassian.net/browse/FW-1635) #resolve

gkalpak committed:
fix(ngcc): do not fail when multiple workers try to create the same directory

When `ngcc` is running in parallel mode (usually when run from the
command line) and the `createNewEntryPointFormats` option is set to true
(e.g. via the `--create-ivy-entry-points` command line option), it can
happen that two workers end up trying to create the same directory at
the same time. This can lead to a race condition, where both check for
the directory existence, see that the directory does not exist and both
try to create it, with the second failing due the directory's having
already been created by the first one. Note that this only affects
directories and not files, because `ngcc` tasks operate on different
sets of files.

This commit avoids this race condition by allowing `FileSystem`'s
`ensureDir()` method to not fail if one of the directories it is trying
to create already exists (and is indeed a directory). This is fine for
the `ensureDir()` method, since it's purpose is to ensure that the
specified directory exists. So, even if the `mkdir()` call failed
(because the directory exists), `ensureDir()` has still completed its
mission.

Related discussion: https://github.com/angular/angular/pull/33049#issuecomment-540485703
FW-1635 #resolve

mary-poppins commented:
You can preview e3cff96 at https://pr33237-e3cff96.ngbuilds.io/.

mary-poppins commented:
You can preview 37d68c9 at https://pr33237-37d68c9.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.ts:
```diff
@@ -36,7 +36,7 @@ export class NodeJSFileSystem implements FileSystem {
     }
     while (parents.length) {
-      this.mkdir(parents.pop() !);
+      this.safeMkdir(parents.pop() !);
```
I was surprised that you put this in a new method, why not make the standard `mkdir()` method handle this scenario?
I can't think of any scenario where we wouldn't want `mkdir()` to be "safe" in this way.

gkalpak replied:
I didn't want to change the semantics of `mkdir`. In both *nix and Node.js `mkdir()` fails if the directory already exists, unless it is called with the `-p` flag on *nix or the `recursive: true` on Node.js (which is only supported since 10.12.0 - [docs][1]).

Our `FileSystem` abstraction does not support the `-p`/`recursive` option in `fs.mkdir()`, so basically `fs.ensureDir()` is our `mkdir -p` equivalent. Thus, I wanted to preserve the behavior that `mkdir` fails when the directory exists but `mkdir -p` (i.e. `ensureDir()`) doesn't.

I could be convinced otherwise, though. WDYT?

[1]: https://nodejs.org/api/fs.html#fs_fs_mkdir_path_options_callback

petebacondarwin replied:
I'm not going to die defending this one ;-)
But it is interesting to note that the mock file-systems have this semantic:

```
mkdir(path: AbsoluteFsPath): void { this.ensureFolders(this._fileTree, this.splitPath(path)); }
```

But I guess you could argue, that if we made `mkdir` have this semantic then there is no point in `ensureDir` :-)

gkalpak replied:
Oh, that looks like a bug in either `MockFileSystem` or `NodeJSFileSystem` (i.e. they are not aligned). I am fine if we decide that `fs.mkdir()` should behave like `mkdir -p` (and drop `ensureDir()`?) as `MockFileSystem` does, but we should be consistent across all `FileSystem` implementations.

gkalpak replied:
Discussed "offline" and decided to get rid of `FileSystem#mkdir()` and keep `FileSystem#ensureDir()` (with the `mkdir -p` semantics).

gkalpak replied:
Added a second commit to remove `mkdir()` (which doesn't seem to be used anyway).

alxhub replied:
👍 I think this is the right resolution.

petebacondarwin reviewed (approved):
Nice work!
Consider making `mkdir` safe by default.

gkalpak committed:
refactor(compiler-cli): remove unused method `FileSystem#mkdir()`

Previously, the `FileSystem` abstraction featured a `mkdir()` method. In
`NodeJSFileSystem` (the default `FileSystem` implementation used in
actual code), the method behaved similar to Node.js' `fs.mkdirSync()`
(i.e. failing if any parent directory is missing or the directory exists
already). In contrast, `MockFileSystem` (which is the basis or mock
`FileSystem` implementations used in tests) implemented `mkdir()` as an
alias to `ensureDir()`, which behaved more like Node.js'
`fs.mkdirSync()` with the `recursive` option set to `true` (i.e.
creating any missing parent directories and succeeding if the directory
exists already).

This commit fixes this inconsistency by removing the `mkdir()` method,
which was not used anyway and only keeping `ensureDir()` (which is
consistent across our different `FileSystem` implementations).

mary-poppins commented:
You can preview cd5e1da at https://pr33237-cd5e1da.ngbuilds.io/.

alxhub commented:
[Presubmit](http://test/OCL:275488908:BASE:275488969:1571416537308:c044b615)
